### DISCUSSION TITLE
Allow API routes to tie into specific methods

### DISCUSSION THREAD

**Discussion by stramel** [vercel/next.js#10583] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

Handling various Methods in API routes is tiresome. If I only want to handle POST, I have to do something like 

```ts
export default async (req: IncomingMessage, res: ServerResponse) => {
  if (req.method !== 'POST') {
    res.status(405).json({
      error: 'Invalid method',
    })
    return
  }

  // Handle POST
}
```

If I want to handle multiple methods I either have to inline them all or break them into separate functions.

## Describe the solution you'd like

I like having the `default` export as a generic request handler. Expanding on that, it would be nice to have named exports for the various Methods supported. So my above example would become,

```ts
export  const post = async (req: IncomingMessage, res: ServerResponse) => {
  // Handle POST
}
```

**Comment by KieraDOG** [comment]
```
import Boom from '@hapi/boom';
import { includes } from 'ramda';

const ALLOWED_METHOD = ['GET', 'HEAD', 'POST', 'PUT', 'DELETE', 'CONNECT', 'OPTIONS', 'TRACE', 'PATCH'];

const withMethod = (method) => (fn) => async (req, res) => {
  const thisMethod = method.toUpperCase();

  if (req.method !== thisMethod || !includes(thisMethod, ALLOWED_METHOD)) {
    const boomified = Boom.methodNotAllowed();
    const { statusCode, payload } = boomified.output;

    res.status(statusCode);
    res.json(payload);

    return;
  }

  await fn(req, res);
};

export default new Proxy(withMethod, {
  get: (target, handler) => target(handler),
});
```

then you can use `withMethod` like this

```
const HealthCheck = (_, res) => {
  res.status(200).send('OK');
};

export default withMethod.GET(HealthCheck);
```

**Comment by stramel** [comment]
Thanks for the very nice code snippet @KieraDOG! I like that setup for a user-land solution, however, how would that implement multiple methods for a single endpoint like GET, POST, PUT for `/api/users`?

**Comment by KieraDOG** [comment]
We dont have a use case for `PUT` and `POST` on same resource, so 

- GET: /pages/api/users/[id].js
- POST: /pages/api/users

**Comment by stramel** [comment]
Ah, I was just throwing out a random example, sorry. I was looking for handling multiple methods like fetching a list of users (GET /users) and adding a new user (POST /users) on a single api file.

**Comment by loteoo** [comment]
I found this tiny helper library to help with this, which works super great with API routes:
https://github.com/jamo/micro-method-router

```javascript
module.exports = methods({
  get: async (req, res) => {
    return res.status(200).end('It's a GET request!');
  },
  post: async (req, res) => {
    return res.status(200).end('It's a POST request!');
  }
})
```


Works well with middlewares too!
```javascript
module.exports = someMiddlewareForAllMethods(methods({
  get: async (req, res) => {
    return res.status(200).end('It's a GET request!');
  },
  post: someMiddleWareJustForPostMethod(async (req, res) => {
    return res.status(200).end('It's a POST request!');
  })
}))
```

But I still think it would be great if Next.js did this out-of-the-box by checking the `get`, `post`, etc exports from the modules.
